onmt/modules/average_attn.py

```diff
@@ -0,0 +1,88 @@
+""" Multi-Head Attention module """
+import math
+import torch
+import torch.nn as nn
+from torch.autograd import Variable
+import numpy as np
+
+from onmt.utils.misc import aeq
+
+from onmt.utils.transformer_util import PositionwiseFeedForward
+
+
+class AverageAttention(nn.Module):
+    """
+    Average Attention module from
+    "Accelerating Neural Transformer via an Average Attention Network"
+    :cite:`https://arxiv.org/abs/1805.00631`.
+
+    Args:
+       model_dim (int): the dimension of keys/values/queries,
+           must be divisible by head_count
+       dropout (float): dropout parameter
+    """
+
+    def __init__(self, model_dim, dropout=0.1):
+        self.model_dim = model_dim
+
+        super(AverageAttention, self).__init__()
+
+        self.average_layer = PositionwiseFeedForward(model_dim, model_dim, dropout)
+        self.gating_layer = nn.Linear(model_dim * 2, model_dim * 2)
+
+
+    def cumulative_average_mask(self, batch_size, inputs_len):
+          """
+          Builds the mask to compute the cumulative average as described in
+          https://arxiv.org/abs/1805.00631 -- Figure 3
+
+          Args:
+            inputs_len: length of the inputs.
+
+          Returns:
+            A Tensor of shape [batch_size, input_len, input_len]
+          """
+
+          triangle = np.tril(np.ones((inputs_len, inputs_len)))
```

Comment:
I think [`torch.tril`](https://pytorch.org/docs/stable/torch.html?highlight=triangular#torch.tril) could be used here. Better to avoid using Numpy if possible.